resolver/resolver.cc

```diff
@@ -1121,6 +1160,189 @@ class ResolveTypeMembersWalk {
+        auto allowTypeMember = true;
+        job.cast->type = TypeSyntax::getResultType(
+            ctx, *job.typeArg, emptySig,
+            TypeSyntaxArgs{allowSelfType, allowRebind, allowTypeMember, core::Symbols::noSymbol()});
+    }
+
+    // Attempts to resolve the type of the given field. Returns `false` if the cast is not yet resolved.
+    static void resolveField(core::MutableContext ctx, ResolveFieldItem &job) {
+        auto cast = job.cast;
+
+        core::SymbolRef scope;
+        auto uid = job.ident;
+        if (uid->kind == ast::UnresolvedIdent::Kind::Class) {
+            if (!ctx.owner.data(ctx)->isClassOrModule()) {
+                if (auto e = ctx.beginError(uid->loc, core::errors::Resolver::InvalidDeclareVariables)) {
+                    e.setHeader("The class variable `{}` must be declared at class scope", uid->name.show(ctx));
+                }
+            }
+
+            scope = ctx.owner.data(ctx)->enclosingClass(ctx);
+        } else {
+            // we need to check nested block counts because we want all fields to be declared on top level of either
+            // class or body, rather then nested in some block
+            if (job.atTopLevel && ctx.owner.data(ctx)->isClassOrModule()) {
+                // Declaring a class instance variable
+            } else if (job.atTopLevel && ctx.owner.data(ctx)->name == core::Names::initialize()) {
+                // Declaring a instance variable
+            } else if (ctx.owner.data(ctx)->isMethod() && ctx.owner.data(ctx)->owner.data(ctx)->isSingletonClass(ctx) &&
+                       !core::Types::isSubType(ctx, core::Types::nilClass(), cast->type)) {
+                // Declaring a class instance variable in a static method
+                if (auto e = ctx.beginError(uid->loc, core::errors::Resolver::InvalidDeclareVariables)) {
+                    e.setHeader("The singleton instance variable `{}` must be declared inside the class body or "
+                                "declared nilable",
+                                uid->name.show(ctx));
+                }
+            } else if (!core::Types::isSubType(ctx, core::Types::nilClass(), cast->type)) {
+                // Inside a method; declaring a normal instance variable
+                if (auto e = ctx.beginError(uid->loc, core::errors::Resolver::InvalidDeclareVariables)) {
+                    e.setHeader("The instance variable `{}` must be declared inside `{}` or declared nilable",
+                                uid->name.show(ctx), "initialize");
+                }
+            }
+            scope = ctx.selfClass();
+        }
+
+        auto prior = scope.data(ctx)->findMember(ctx, uid->name);
+        if (prior.exists()) {
+            if (core::Types::equiv(ctx, prior.data(ctx)->resultType, cast->type)) {
+                // We already have a symbol for this field, and it matches what we already saw, so we can short
+                // circuit.
+                return;
```

Comment:
At this point in the code, we've discovered that there's a field with the same name (it has to be a field since `uid->name` is going to be `@something`). If the field is defined, it must have a `resultType` (we never define a field and _not_ add `resultType`). So now it's checking if the type matches.